### PR TITLE
use AlertmanagerConfigAllowList for alertmanager secret informer

### LINKED ISSUE #3544: Support and document tighter RBAC permissions for Operator
Some of the RBAC permissions that must be granted to the Operator, in its ClusterRole, seem too broad.  Following the principle of least privilege's, can they be tightened up?  If so, how?

**Details**

The existing permissions in bundle.yaml seem to fall into the following groups:

|Area|Example|Comment on Security|
|---|---|---|
|Access to CR instances|e.g. alertmanagers, serivcemonitors etc in API group monitoring.coreos.com|This seems fine to me, as is. Since the Operator is only messing with its own stuff|
|get/list/watch for scraping purposes|e.g. find pods, endpoints etc|Also seems fine. Is readonly and part of what the Operator is meant to do|
|modifications to services and endpoints, delete of pods||It would be **great if this could be scoped by namespace**, so that the Operator is only able to mess with stuff in a designated namespace (the one where it installs Prometheus)|
|* (all verbs) on configmaps, secrets and statefulsets||Again, it would be **great if this could be scoped down to a single namespace**, so that the Operator is provably only messing with its _own_ configmaps, secrets and statefulsets|

Would this be possible?  Could guidance be published on how to do it?

BTW, I'm guessing that maybe one approach would be to use a ClusterRole for the stuff that's fine as is, and a Role (single namespace) for the ones that we'd like to tighten up.

I think this request is related, but not identical, to https://github.com/prometheus-operator/prometheus-operator/issues/3543

### PR DESCRIPTION
**Pull request by william-tran**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

I want to use the operator to handle PrometheusRule CRDs in many namespaces but only handle Prometheus CRDs in one namespace.  I saw the default RBAC in the helm chart was far too broad and wrote my own, having secrets operations as Roles instead of ClusterRoles, attached only to the namespaces where I want to set up prometheus instances.  I noticed after setting `--namespaces` and `--prometheus-instance-selector` to different values, that list/watch of secrets was being attempted in the namespaces under `--namespaces`, with access denied.

Fixes #3544 (and #3298 in the case `--namespaces` is unset, which results in attempting to watch all secrets)

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Added `-alertmanager-config-namespaces` CLI flag to the operator. 
```

### CODE REVIEW

**Comment by simonpasquier** [comment]
This secrets informer exists for `AlertmanagerConfig` custom resources which themselves can reference secret keys (for defining receiver's credentials for instance). These resources live in the namespaces defined by `--namespaces` and not `--alertmanager-instance-namespaces` which is why the informer is set up this way.
Now the operator should probably have another secrets informer watching for `alertmanager-instance-namespaces` to reconcile changes to the base Alertmanager configuration's secret...
cc @PhilipGough

**Comment by william-tran** [comment]
> These resources live in the namespaces defined by --namespaces
Ahh I see now with the `Alertmanager` CRD vs `AlertmanagerConfig` which can reference secrets e.g. https://prometheus-operator.dev/docs/operator/api/#slackconfig `apiURL`

To address tighter scoping of secrets permissions, we could add an option alertmanager-config-namespaces. This could cater to those who setup alertmanager config centrally versus having anyone provide that config.

**Comment by simonpasquier** [comment]
> To address tighter scoping of secrets permissions, we could add an option alertmanager-config-namespaces

I wouldn't be opposed to this. To be consistent with the other flags, the logic should be:
* use `--alertmanager-config-namespaces` if not empty.
* otherwise use `--namespaces` (as before).

@fpetkovski @paulfantom thoughts?

**Comment by PhilipGough** [comment]
@simonpasquier  Sounds reasonable to me. Will wait for others to chime in.

**Comment by fpetkovski** [comment]
This suggestion makes sense to me as well :+1:

**Comment by paulfantom** [comment]
Simon's suggestion sounds good to me.

**Comment by william-tran** [comment]
Great, I'll get to work on this and will have something by end of week.

File: pkg/alertmanager/operator.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ func (c *Operator) bootstrap(ctx context.Context) error {
 	}
 	c.secrInfs, err = informers.NewInformersForResource(
 		informers.NewKubeInformerFactories(
-			c.config.Namespaces.AllowList,
+			c.config.Namespaces.AlertmanagerConfigAllowList,
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:191]: L171 needs to be modified too?

**Review by simonpasquier — APPROVED** [review]
lgtm

**Comment by fpetkovski** [comment]
@william-tran some CI jobs are failing, you should be able to fix them with `make format generate`.

**Review by fpetkovski — APPROVED** [review]

**Comment by fpetkovski** [comment]
@simonpasquier okay with you to merge?

**Comment by william-tran** [comment]
The check docs error seems unrelated to this PR:

https://github.com/prometheus-operator/prometheus-operator/runs/5004881958?check_suite_focus=true#step:5:254

```
error: 		ADOPTERS.md:32: "https://www.clyso.com/en" not accessible even after retry; status code 0: Get "https://www.clyso.com/en": unexpected EOF
```

**Comment by PhilipGough** [comment]
@william-tran - if you rebase on top of `main` that check should resolve. It is not required however so we can merge as is.

**Comment by william-tran** [comment]
@PhilipGough done

**Review by fpetkovski — APPROVED** [review]

**Comment by simonpasquier** [comment]
thanks!

**Pull request merged**